Ability to use a stream to write into a NpgsqlWriteBuffer

YohDeadfall opened the pull request:
Does it make sense to move streams into the buffer classes and make them nested in additional `NpgsqlXxxBuffer.Stream.cs` files? Also I would like to add argument validation in the `ColumnStream` class, so do not merge.

Fixes #1868

roji commented:
> Does it make sense to move streams into the buffer classes and make them nested in additional NpgsqlXxxBuffer.Stream.cs files?

Sure, why not!

> Also I would like to add argument validation in the ColumnStream class, so do not merge.

OK, let me know when to review.

YohDeadfall commented on src/Npgsql/ColumnStream.cs:
```diff
@@ -1,162 +0,0 @@
-
-        public override Task<int> ReadAsync([NotNull] byte[] buffer, int offset, int count, CancellationToken token)
-        {
-            token.ThrowIfCancellationRequested();
```
This method shouldn't be used in non-async methods. Fixed.

YohDeadfall commented:
Things which were done:

* Moved classes.
* Added argument validation.
* Optimized the `WriteAsync` method in `ParameterStream`.
* Fixed an issue in `ColumnStream`.

Yay, you fixed AppVeyor! But now Travis CI fails.

roji commented on src/Npgsql/NpgsqlWriteBuffer.Stream.cs:
```diff
@@ -0,0 +1,146 @@
+{
+    public sealed partial class NpgsqlWriteBuffer
+    {
+        sealed class ParameterStream : Stream
```
I don't like the name very much... Why not have `ColumnStream` just like on the read side - each is internal to its buffer class so it's OK? Or we could have `ColumnReadStream` and `ColumnWriteStream`.

roji commented on src/Npgsql.NetTopologySuite/NetTopologySuiteHandler.cs:
```diff
@@ -274,74 +274,14 @@ Task INpgsqlTypeHandler<GeometryCollection>.Write(GeometryCollection value, Npgs
-            _writer.Write(value, new WriteStream(buf));
+            _writer.Write(value, buf.GetStream());
 #if NET45
             return Task.Delay(0);
```
I know this isn't a part of this PR, but I just noticed it. This `#if` can be eliminated simply by returning `PGUtil.CompletedTask` which works on all target frameworks.

roji commented:
BTW I restarted the Travis build and everything's green now, hurray...
